lib/auth/accessmonitoring/accessmonitoring.go

```diff
@@ -0,0 +1,112 @@
+
+	accessmonitoringrulesv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessmonitoringrules/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/accessmonitoring"
+	"github.com/gravitational/teleport/lib/accessmonitoring/review"
+	"github.com/gravitational/teleport/lib/backend"
+)
+
+// Client aggregates the parts of Teleport API client interface
+// (as implemented by github.com/gravitational/teleport/api/client.Client)
+// that are used by the access plugins.
+type Client interface {
+	types.Events
+	SubmitAccessReview(ctx context.Context, params types.AccessReviewSubmission) (types.AccessRequest, error)
+	ListAccessMonitoringRulesWithFilter(ctx context.Context, req *accessmonitoringrulesv1.ListAccessMonitoringRulesWithFilterRequest) ([]*accessmonitoringrulesv1.AccessMonitoringRule, string, error)
+	GetUser(ctx context.Context, name string, withSecrets bool) (types.User, error)
+}
+
+// Config specifies the access monitoring service configuration.
+type Config struct {
+	// Logger is the logger for the access monitoring serivce.
+	Logger *slog.Logger
+
+	// Backend should be a backend.Backend which can be used for obtaining the
+	// lock required to run the service.
+	Backend backend.Backend
+
+	// Client is the auth service client interface.
+	Client Client
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Logger == nil {
+		c.Logger = slog.Default()
+	}
+	if c.Backend == nil {
+		return trace.BadParameter("backend: must be non-nil")
+	}
+	if c.Client == nil {
+		return trace.BadParameter("client: must be non-nil")
+	}
+	return nil
+}
+
+// AccessMonitoringService monitors access events and applies access monitoring
+// rules.
+type AccessMonitoringService struct {
+	cfg Config
+}
+
+// NewAccessMonitoringSerivce returns a new access monitoring service.
+func NewAccessMonitoringService(cfg Config) (*AccessMonitoringService, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err, "failed to validate access monitoring service config")
+	}
+	return &AccessMonitoringService{
+		cfg: cfg,
+	}, nil
+}
+
+// Run the access monitoring service.
+func (s *AccessMonitoringService) Run(ctx context.Context) (err error) {
+	accessReviewHandler, err := review.NewHandler(review.Config{
+		Logger:      s.cfg.Logger,
+		HandlerName: types.BuiltInAutomaticReview,
+		Client:      s.cfg.Client,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	monitor, err := accessmonitoring.NewAccessMonitor(accessmonitoring.Config{
+		Logger:  s.cfg.Logger,
+		Backend: s.cfg.Backend,
+		Events:  s.cfg.Client,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Configure access review handlers.
+	monitor.AddAccessMonitoringRuleHandler(accessReviewHandler.HandleAccessMonitoringRule)
+	monitor.AddAccessRequestHandler(accessReviewHandler.HandleAccessRequest)
+
+	return trace.Wrap(monitor.Run(ctx))
+}
```

Comment:
In the unfortunate event of the accessmonitoring service panicking, can we defer a `recover` to avoid bringing the whole Teleport auth down and triggering an outage?

There's an example here: https://github.com/gravitational/teleport/blob/c5999a83720c24688f7d08c0a7b471296796e034/lib/autoupdate/rollout/controller.go#L136-L155

Reply:
Good call. Updated in https://github.com/gravitational/teleport/pull/54000/commits/d349bb79cd74e08cf6554155de9f21fb37f1524c